File: TripSync-AI-Trip-Planner/services/dining.service.js
import axios from 'axios';

export const getRestaurants = async (lat, lon, radius = 1000) => {
    const overpassUrl = "https://overpass-api.de/api/interpreter";
    const query = `
    [out:json][timeout:5];
    (
      node["amenity"="restaurant"](around:${radius},${lat},${lon});
      way["amenity"="restaurant"](around:${radius},${lat},${lon});
    );
    out center 15;
  `;

    try {
        console.log("Fetching dining data...");
        const response = await axios.post(overpassUrl, `data=${encodeURIComponent(query)}`, {
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            timeout: 8000 // Increased timeout to 8s
        });

        const restaurants = [];

        // Check for valid response structure
        if (!response.data || !response.data.elements) {
            console.log("No dining data received");
            return [];
        }

        for (const element of response.data.elements) {
            const tags = element.tags || {};
            const name = tags.name;
            const cuisine = tags.cuisine;
            const city = tags["addr:city"];
            const address = tags["addr:full"] || tags["addr:street"] ? `${tags["addr:number"] || ''} ${tags["addr:street"] || ''}`.trim() : "";

            if (name) {
                restaurants.push({
                    name: name,
                    cuisine: cuisine || "Various",
                    city: city || "",
                    address: address || "",
                    type: "restaurant",
                    lat: element.lat || element.center?.lat,
                    lon: element.lon || element.center?.lon
                });
            }
        }

        // Return top 10 unique restaurants
        return restaurants
            .filter((v, i, a) => a.findIndex(t => (t.name === v.name)) === i)
            .slice(0, 10);

    } catch (error) {
        console.error("Error fetching restaurants:", error.message);
        return [];
    }
};

File: services/dining_service.py
"""
Dining Service - Fetches restaurants using Overpass API
"""
import requests
import logging

logger = logging.getLogger(__name__)


def get_restaurants(lat: float, lon: float, radius: int = 1000) -> list:
    """
    Fetch restaurants near coordinates using Overpass API.
    
    Args:
        lat: Latitude
        lon: Longitude
        radius: Search radius in meters (default 1000)
    
    Returns:
        List of restaurant dicts with name, cuisine, address, lat, lon
    """
    overpass_url = "https://overpass-api.de/api/interpreter"
    query = f"""
    [out:json][timeout:5];
    (
      node["amenity"="restaurant"](around:{radius},{lat},{lon});
      way["amenity"="restaurant"](around:{radius},{lat},{lon});
    );
    out center 15;
    """

    try:
        logger.info("Fetching dining data...")
        response = requests.post(
            overpass_url,
            data={"data": query},
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            timeout=8
        )
        response.raise_for_status()
        data = response.json()

        restaurants = []
        elements = data.get("elements", [])

        for element in elements:
            tags = element.get("tags", {})
            name = tags.get("name")

            if not name:
                continue

            cuisine = tags.get("cuisine", "Various")
            city = tags.get("addr:city", "")
            
            # Build address
            addr_number = tags.get("addr:number", "")
            addr_street = tags.get("addr:street", "")
            address = f"{addr_number} {addr_street}".strip() if addr_street else ""

            # Get coordinates (handle both node and way elements)
            elem_lat = element.get("lat") or (element.get("center", {}).get("lat"))
            elem_lon = element.get("lon") or (element.get("center", {}).get("lon"))

            restaurants.append({
                "name": name,
                "cuisine": cuisine,
                "city": city,
                "address": address,
                "type": "restaurant",
                "lat": elem_lat,
                "lon": elem_lon
            })

        # Return top 10 unique restaurants
        seen_names = set()
        unique_restaurants = []
        for r in restaurants:
            if r["name"] not in seen_names:
                seen_names.add(r["name"])
                unique_restaurants.append(r)
                if len(unique_restaurants) >= 10:
                    break

        return unique_restaurants

    except requests.RequestException as e:
        logger.error(f"Error fetching restaurants: {e}")
        return []
